common/expression/src/function.rs

```diff
@@ -0,0 +1,691 @@
+                        .filter_map(|(id, func)| {
+                            if func.signature.name == *name
+                                && func.signature.args_type.len() == args_type.len()
+                            {
+                                Some((FunctionID::Builtin { name, id }, func.clone()))
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<_>>()
+                })
+                .unwrap_or_default();
+
+            if !builtin_funcs.is_empty() {
+                return builtin_funcs;
+            }
+        }
+
+        self.factories
+            .get_key_value(name)
+            .map(|(name, factories)| {
+                factories
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(id, factory)| {
+                        factory(params, args_type).map(|func| {
+                            (
+                                FunctionID::Factory {
+                                    name,
+                                    id,
+                                    params: params.to_vec(),
+                                    args_type: args_type.to_vec(),
+                                },
+                                func,
+                            )
+                        })
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .unwrap_or_default()
+    }
+
+    pub fn register_0_arg_core<O: ArgType, F>(
+        &mut self,
+        name: &'static str,
+        property: FunctionProperty,
+        func: F,
+    ) where
+        F: Fn(&GenericMap) -> Value<O> + 'static + Clone + Copy,
+    {
+        self.funcs
+            .entry(name)
+            .or_insert_with(Vec::new)
+            .push(Arc::new(Function {
+                signature: FunctionSignature {
+                    name,
+                    args_type: vec![],
+                    return_type: O::data_type(),
+                    property,
+                },
+                eval: Box::new(erase_function_generic_0_arg(func)),
+            }));
+    }
+
+    pub fn register_1_arg<I1: ArgType, O: ArgType, F>(
+        &mut self,
+        name: &'static str,
+        property: FunctionProperty,
+        func: F,
+    ) where
+        F: for<'a> Fn(I1::ScalarRef<'a>) -> O::Scalar + 'static + Clone + Copy,
+    {
+        let has_nullable = &[I1::data_type(), O::data_type()]
+            .iter()
+            .any(|ty| ty.as_nullable().is_some());
+
+        assert!(
+            !has_nullable,
+            "Function {} has nullable argument or output, please use register_1_arg_core instead",
+            name
+        );
+
+        let property = property.preserve_not_null(true);
+
+        self.register_1_arg_core::<NullType, NullType, _>(name, property.clone(), move |_, _| {
+            Value::Scalar(())
```

Comment:
Maybe we should keep column length if it is a column?

Reply:
Is seems not necessary to keep a column length for a scalar. Is there a place we had used?

Reply:
Thanks, got it, NullType is just an inner data type.

Reply:
NullType is not a inner type, but similar to `Nullable<Nothing>` in clickhouse.

Reply:
I think it still need column length here. It is not clear in 1 arg function.

For example, we have a 2-arg-function `plus`, the inputs are `Literal::Null` and`ColumnRef(UInt8)`.

The type check will select `plus(Null, UInt8)` as the finaly function.

The result will be a scalar, so it is not correct.